### PR TITLE
Feat: enable `securityContext` on spire server pod

### LINKED ISSUE #27130: SPIRE server missing correct SELinux context and capabilities options
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

1. Fedora 38
2. deploy Cilium using Helm chart with SPIRE server enabled
3. SPIRE server won't bind socket within `/run` on host because SELinux context is not set properly

### Cilium Version

`1.41.0`

### Kernel Version

`6.4.4-200.fc38.x86_64`

### Kubernetes Version

`1.27.4`

### Sysdump

*N/A*

### Relevant log output

Relevant excerpt from `/var/log/audit/audit.log`:

```yaml
type=AVC msg=audit(1690532456.762:87318): avc:  denied  { write } for  pid=365463 comm="spire-server" name="sockets" dev="tmpfs" ino=43170 scontext=system_u:system_r:container_t:s0:c375,c775 tcontext=system_u:object_r:container_var_run_t:s0 tclass=dir permissive=0
```

### Anything else?

I tried to add the following context as quick fix to the Helm chart for the SPIRE server container:

```yaml
specs:
  containers:
  -
    securityContext:
      seLinuxOptions:
        level: s0
        type: spc_t
```

It effectively resolved the issue and SPIRE server binded the socket.

Also add relevant `securityContext.capabilities` for those enforcing strict policies, please. Thank you.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by ishuar**
Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [ ] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [ ] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [ ] Thanks for contributing!

- Fixes: https://github.com/cilium/cilium/issues/27130

```release-note
Add securityContext for spire pod in helm chart
```

### CODE REVIEW

**Review by nebril — CHANGES_REQUESTED** [review]
Thanks for the PR! The change itself looks good. I think the PR should be squashed into one commit though, please do that and re-request review when you have repushed branch, I will run CI then.

**Comment by ishuar** [comment]
> Thanks for the PR! The change itself looks good. I think the PR should be squashed into one commit though, please do that and re-request review when you have repushed branch, I will run CI then.

Thank you for your response, I have squashed my commits and pushed it. 

Just curious if we could also use squash and commit while merging this PR in case this is needed for the future too?

**Review by nebril — CHANGES_REQUESTED** [review]
@ishuar we don't use squash and merge from GH because we usually want PRs to be structured logically. If the change is big, we want to have multiple commits, but your change simply seemed too fragmented. So we end up rebasing most PRs to have clean git history.

I can still see merge commit in your PR, can you get rid of it?

File: install/kubernetes/cilium/templates/spire/server/statefulset.yaml
Comment (qmonnet) [inline install/kubernetes/cilium/templates/spire/server/statefulset.yaml]: Nit: there's a double space between `toYaml` and `.` on this line.

**Review by qmonnet — CHANGES_REQUESTED** [review]
Please run `make -C Documentation update-helm-values` from the root of the repository to update the Helm reference in the docs, and add the changes to your commit.

Looks good to me otherwise.

**Review by qmonnet — COMMENTED** [review]
Please also rework your commit description:

- It looks like you haven't updated after squashing commits and it contains logs and titles from multiple commits.
- It also contains the logs of one unrelated commit, with a wrong `Signed-off-by` tag from myself, I guess because of your rebasing.

File: install/kubernetes/cilium/templates/spire/server/statefulset.yaml
Comment (ishuar) [reply install/kubernetes/cilium/templates/spire/server/statefulset.yaml]: Thanks, Done.  I am not sure why `:spelling:ignore:` is removed from `helm-values.rst`

**Comment by ishuar** [comment]
> Please also rework your commit description:
> 
> * It looks like you haven't updated after squashing commits and it contains logs and titles from multiple commits.
> * It also contains the logs of one unrelated commit, with a wrong `Signed-off-by` tag from myself, I guess because of your rebasing.

Also fixed

File: install/kubernetes/cilium/templates/spire/server/statefulset.yaml
Comment (qmonnet) [reply install/kubernetes/cilium/templates/spire/server/statefulset.yaml]: This is likely because your `main` branch is too old, and the Documentation Makefile does not contain the code to add these tags yet? Can you try to rebase your PR on the top of the current main (`git checkout main; git rebase main feat/enable-podsecuritycontext-on-spire-server`) and re-run the command, please?

**Review by qmonnet — COMMENTED** [review]
Documentation workflow is failing on spell check, this is because you unintentionally removed the `:spelling:ignore:` tags from the Helm ref, see [my comment here](https://github.com/cilium/cilium/pull/27363#discussion_r1293330709).

Checkpatch complains that your commit title is too long, but in fact this is just that you're missing an empty line between the commit title and the body of the commit description (GitHub considers only the first line as the title, but without empty-line separator, other tools usually consider the next lines as the continuation of a wrapped commit message).

File: install/kubernetes/cilium/templates/spire/server/statefulset.yaml
Comment (ishuar) [reply install/kubernetes/cilium/templates/spire/server/statefulset.yaml]: I have tried the following and yet not getting `:spelling:ignore:` prefix.


 
 1. Updated/synced my fork at https://github.com/ishuar/cilium/tree/main
 2.  Checkout main locally and ran `git pull`  , (latest commit is in the rebase output)
 3.  Rebase the branch `git rebase main feat/enable-podsecuritycontext-on-spire-server`
 4. ran `make -C Documentation update-helm-values` from the the repo root path.[ `<PATH_TO_CILLIUM_FORKED_REPO>/cilium`]

- Bash Version if helps,
```bash
╰─$ bash --version              
GNU bash, version 3.2.57(1)-release (arm64-apple-darwin22)
Copyright (C) 2007 Free Software Foundation, Inc.
```

>> in my forked repo [Documentation/makefile](https://github.com/ishuar/cilium/blob/main/Documentation/Makefile#L113)  `:spelling:ignore:`  is available atleast.

<details>
<summary>Click to check the local commands workflow</summary>


```bash
╰─$ git checkout main ; git pull ; git rebase main feat/enable-podsecuritycontext-on-spire-server
Switched to branch 'main'
Your branch is up to date with 'origin/main'.
Already up to date.
Successfully rebased and updated refs/heads/feat/enable-podsecuritycontext-on-spire-server.
╰─$ make -C Documentation update-helm-values
docker container run --rm --workdir /src/Documentation --volume <PATH_TO_CILLIUM_FORKED_REPO>cilium/Documentation/..:/src --env READTHEDOCS_VERSION= --env SKIP_LINT= --user "501:20" "quay.io/cilium/helm-toolbox:"v1.0.1"@sha256:"a1483fc036b15c290326ba7ed36b8f0b714e185913b72a6074c04225d74e034c"" helm-docs -d -c /src/install/kubernetes -t /src/Documentation/helm-values.tmp.tmpl > helm-values.tmp
time="2023-08-14T13:53:14Z" level=info msg="Found Chart directories [cilium]"
time="2023-08-14T13:53:14Z" level=info msg="Generating README Documentation for chart /src/install/kubernetes/cilium"
awk -F'|' '{print "|"$2"|"$5"|"$3"|"$4"|"}' helm-values.tmp > helm-values.awk
docker container run --rm --workdir /src/Documentation --volume /Users/is/Documents/Personal-GitHub/cilium/Documentation/..:/src --env READTHEDOCS_VERSION= --env SKIP_LINT= --user "501:20" "quay.io/cilium/helm-toolbox:"v1.0.1"@sha256:"a1483fc036b15c290326ba7ed36b8f0b714e185913b72a6074c04225d74e034c"" python3 /usr/bin/m2r2 --overwrite helm-values.awk
sed 's/^\(   \* - \)\([[:print:]]\+\)$/\1:spelling:ignore:`\2`/' helm-values.rst > helm-values.sed
printf '..\n  %s\n\n%s\n' "AUTO-GENERATED. Please DO NOT edit manually." "$(cat helm-values.sed)" > helm-values.rst
rm -f -- helm-values.tmp helm-values.awk helm-values.sed
╰─$ git log --oneline --max-count=10                                                             
ed30c230fc (HEAD -> feat/enable-podsecuritycontext-on-spire-server) added the securityContext in Spire server pod and container.
f439a14d8e (origin/main, origin/HEAD, main) bpf: policy: fine-tune logic in __policy_can_access()
b113273df2 bpf: policy: fine-tune ICMP logic
ef7f6f4112 bpf: policy: add policy_can_ingress*() wrappers
1410a661c1 Add SPIRE connection to status
916a2ce2c2 health: only launch /hello after host datapath is ready
6c0d4787f0 Bump Golang to v1.20 on Cilium v1.12 and v1.13
92481c499b egressgw: back out test for policy conflict in ENI mode
3c366bd3b9 docs, cilium: Remove `cilium endpoint regenerate` command
c49bf315e6 hive: prevent goleak error due to race condition

```

</details>

Comment (qmonnet) [reply install/kubernetes/cilium/templates/spire/server/statefulset.yaml]: @ishuar 

```
sed 's/^\(   \* - \)\([[:print:]]\+\)$/\1:spelling:ignore:`\2`/' helm-values.rst > helm-values.sed
printf '..\n  %s\n\n%s\n' "AUTO-GENERATED. Please DO NOT edit manually." "$(cat helm-values.sed)" > helm-values.rst
```

These are the commands that should take care of adding the tags indeed. Did you forget to  `git add` and `git commit` the changes after that? Or is the file still missing the role tags after you ran this command?

Comment (ishuar) [reply install/kubernetes/cilium/templates/spire/server/statefulset.yaml]: @qmonnet I also have the same understanding of the commands. but seems like sed on macOS does not work as expected in the make command. 

I have the correct makefile locally as you can see in my execution it is running `sed 's/^\(   \* - \)\([[:print:]]\+\)$/\1:spelling:ignore:`\2`/' helm-values.rst > helm-values.sed` but not inserting the string `:spelling:ignore:` to the regex pattern.

Comment (qmonnet) [reply install/kubernetes/cilium/templates/spire/server/statefulset.yaml]: Oooh, macOS, I get it now. I tried to fix this command once for macOS already, but I don't have any Apple hardware around to test it, so I take it my fix was not good enough.

Would you mind helping me figure out the right command? I wonder if this is the `[[:print:]]` which is not supported on your version. What do you get with:

```
$ echo foo | sed 's/[[:print:]]/b/'    
boo
```

Alternatively, if you have GNU sed installed (`gsed`), the Makefile should pick it up over your regular `sed` and it should work. But it would be great to fix this command once and for all.

Comment (ishuar) [reply install/kubernetes/cilium/templates/spire/server/statefulset.yaml]: I installed `gnu-sed` for verifying and yes it's the default `sed` in MacOS causing the issue.

```bash
╰─$ gsed  's/^\(   \* - \)\([[:print:]]\+\)$/\1:spelling:ignore:`\2`/' helm-values.rst > helm-values.sed | grep -i "spelling" | head -n 5
   * - :spelling:ignore:`Key`
   * - :spelling:ignore:`MTU`
   * - :spelling:ignore:`affinity`
   * - :spelling:ignore:`agent`
   * - :spelling:ignore:`agentNotReadyTaintKey`
  ```
  
  ```bash
─$ gsed --version                                                                                                                       
gsed (GNU sed) 4.9
```

Comment (qmonnet) [reply install/kubernetes/cilium/templates/spire/server/statefulset.yaml]: From https://riptutorial.com/sed/topic/9436/bsd-macos-sed-vs--gnu-sed-vs--the-posix-sed-specification, it's maybe the `\+` that causes the issue. We could try to replace it with `\{1,\}`, or simply with `*` (as I don't expect the value name field to be empty).

Comment (ishuar) [reply install/kubernetes/cilium/templates/spire/server/statefulset.yaml]: I get the same output
```bash
$ echo foo | sed 's/[[:print:]]/b/'  
boo
```

Comment (ishuar) [reply install/kubernetes/cilium/templates/spire/server/statefulset.yaml]:
```bash
sed 's/^\(   \* - \)\([[:print:]]\{1,\}\)$/\1:spelling:ignore:`\2`/' 
```
 worked 🎆

Comment (ishuar) [reply install/kubernetes/cilium/templates/spire/server/statefulset.yaml]: @qmonnet the above command is working fine on macOs with default `sed` may be you could verify on other platforms if required.

Also, do you think I should create another PR for the helm docs fix for macOS or can it be part of this PR?

Comment (qmonnet) [reply install/kubernetes/cilium/templates/spire/server/statefulset.yaml]: Great! Thanks for testing

> do you think I should create another PR for the helm docs fix for macOS or can it be part of this PR?

Ideally another PR, because we'll need to mark it for backport to all stable branches (unlike this PR).

Comment (ishuar) [reply install/kubernetes/cilium/templates/spire/server/statefulset.yaml]: @qmonnet Thanks for your time and support , here is the new PR 

- https://github.com/cilium/cilium/pull/27495

Comment (qmonnet) [reply install/kubernetes/cilium/templates/spire/server/statefulset.yaml]: Thanks a lot! I'll take a look tomorrow.

**Review by qmonnet — COMMENTED** [review]
Now the Helm reference is fixed, great!

This time the CI complains because of `seccompProfile`, which is used in your description, but unknown to the spell checker.

Can you please run `./Documentation/update-spelling_wordlist.sh seccompProfile`, and add the change to your commit?

File: install/kubernetes/cilium/templates/spire/server/statefulset.yaml
Comment (qmonnet) [reply install/kubernetes/cilium/templates/spire/server/statefulset.yaml]: Now merged :+1:

**Review by qmonnet — APPROVED** [review]
Looks good from my side, thank you!

**Comment by ishuar** [comment]
> @ishuar we don't use squash and merge from GH because we usually want PRs to be structured logically. If the change is big, we want to have multiple commits, but your change simply seemed too fragmented. So we end up rebasing most PRs to have clean git history.
> 
> I can still see merge commit in your PR, can you get rid of it?

@nebril could you please re-review, I do not have any option to dismiss the changes you have requested. 

or @qmonnet  would it be fine for you to dismiss the change request at [here](https://github.com/cilium/cilium/pull/27363#pullrequestreview-1573134750)

**Comment by qmonnet** [comment]
> @qmonnet would it be fine for you to dismiss the change request at [here](https://github.com/cilium/cilium/pull/27363#pullrequestreview-1573134750)

We need approvals from all codeowner teams before merging, so I'd rather wait a bit longer and have Maciej's and Alexandre's reviews (ping @kaworu). I'm not familiar with SPIRE myself, so I won't bypass the other teams here.

**Comment by qmonnet** [comment]
End-to-end tests in CI fail consistently, but I'm unsure this is related to your PR (I've read that the recent switch to go 1.21 could cause this, not sure if this is what we're hitting).

Could you please rebase on top of the current `main` branch? Sorry for the inconvenience

**Review by nebril — APPROVED** [review]
LGTM, thanks for fixing all the things in the process! I think a rebase on main might help with the CI, but I am approving anyway, it should be retained over the rebase.

**Comment by qmonnet** [comment]
@ishuar Sorry, I'll ask you to rebase once more :disappointed:. There's a conflict with https://github.com/cilium/cilium/pull/27229, which was just merged.

**Review by kaworu — CHANGES_REQUESTED** [review]
Not sure why @cilium/sig-hubble was required for review, probably an older revision of the patch. Helm changes LGTM, pulling in @mhofstetter to take a look at the spire stuff.

**Review by kaworu — APPROVED** [review]
wrong button :sweat_smile:

**Comment by ishuar** [comment]
> @ishuar Sorry, I'll ask you to rebase once more 😞. There's a conflict with #27229, which was just merged.

hopefully, this should be fine...

File: install/kubernetes/cilium/values.yaml.tmpl
Comment (mhofstetter) [inline install/kubernetes/cilium/values.yaml.tmpl]: having `runAsNonRoot` set to `true` results in the following error running the SPIRE server locally on kind.

```
container has runAsNonRoot and image will run as root (pod: "spire-server-0_cilium-spire(36e48818-7db4-4a4e-87a1-2ac6e5a5a27d)", container: spire-server)
```

Let's wait what the tests on CI report.

**Review by mhofstetter — CHANGES_REQUESTED** [review]
I get an error running the SPIRE server locally on kind (see inline comment).

We might want to change the default?

File: install/kubernetes/cilium/values.yaml.tmpl
Comment (mhofstetter) [reply install/kubernetes/cilium/values.yaml.tmpl]: Looks like CI is failing with the same issue

e.g. https://github.com/cilium/cilium/actions/runs/5889014601 (CI-E2E)

`cilium-spire         spire-server-0                               1/2     CreateContainerConfigError `

```
status:
...
    - image: ghcr.io/spiffe/spire-server:1.6.3@sha256:f4bc49fb0bd1d817a6c46204cc7ce943c73fb0a5496a78e0e4dc20c9a816ad7f
      imageID: ""
      lastState: {}
      name: spire-server
      ready: false
      restartCount: 0
      started: false
      state:
        waiting:
          message: 'container has runAsNonRoot and image will run as root (pod: "spire-server-0_cilium-spire(6938b1d5-0598-42ab-a23e-5af71f62f7fc)",
            container: spire-server)'
          reason: CreateContainerConfigError
```

Comment (mhofstetter) [reply install/kubernetes/cilium/values.yaml.tmpl]: I'd recommend to either changing the default to the recommended approach in the [related issue](https://github.com/cilium/cilium/issues/27130) (i haven't tested this yet locally) or leaving it empty (and therefore up to the user configuring it to its needs)

Comment (ishuar) [reply install/kubernetes/cilium/values.yaml.tmpl]: Done, removed the defaults. now it's flexible for the user.

File: Documentation/spelling_wordlist.txt
Comment (mhofstetter) [inline Documentation/spelling_wordlist.txt]: @ishuar with the removal of the initial defaults in the `SecurityContext`, this entry in the spelling wordlist is no longer necessary

File: install/kubernetes/cilium/values.yaml.tmpl
Comment (mhofstetter) [reply install/kubernetes/cilium/values.yaml.tmpl]: thanks!

File: Documentation/spelling_wordlist.txt
Comment (ishuar) [reply Documentation/spelling_wordlist.txt]: @mhofstetter  removed

**Review by meyskens — APPROVED** [review]
LGTM for me

**Review by mhofstetter — APPROVED** [review]
LGTM - thanks for the contribution.

**Comment by mhofstetter** [comment]
Check "Conformance GatewayAPI" fails on `main` too and will be fixed with https://github.com/cilium/cilium/pull/27592

Marking this PR as `ready-to-merge`

**Pull request merged**
